azkaban-common/src/main/java/azkaban/flow/FlowUtils.java

```diff
@@ -104,10 +106,27 @@ public static Flow getFlow(final Project project, final String flowName) {
     return flow;
   }
 
-  public static ExecutableFlow createExecutableFlow(final Project project, final Flow flow) {
-    final ExecutableFlow exflow = new ExecutableFlow(project, flow);
-    exflow.addAllProxyUsers(project.getProxyUsers());
-    return exflow;
+  // if pass in executorManagerAdapter, then use it to populate the flow's ExecutionOption
+  public static ExecutableFlow createExecutableFlow(final Project project, final Flow flow,
+   ExecutorManagerAdapter executorManagerAdapter, Logger logger) {
+    ExecutableFlow exFlow = new ExecutableFlow(project, flow);
+    exFlow.addAllProxyUsers(project.getProxyUsers());
+
+    if (executorManagerAdapter == null) {
+      return exFlow;
+    }
+    try {
+      executorManagerAdapter.preloadExecutionOptions(exFlow);
```

Comment:
This way we are going back to the start.
Separate `fetching Props from project` (IO thing) from `setting Props into ExecutableFlow`(only in memory) is the key here.

Reply:
ah okay I think I get the point now.

Reply:
How about taking out createExecutableFlow from util to a method of ExecutorManagerAdapter? 
I think it is a class method instead of util anymore b/c a well-defined ExecutorManagerAdapter class object is passed in

Reply:
good point!